text/3366-diagnostic-attribute-namespace.md

```diff
@@ -0,0 +1,222 @@
+    note = "That's emitted if A and Self is a i32"
+)]
+#[diagnostic::on_unimplemented(
+    if(not(A = String)),
+    // and implicitly `A` is not a `i32` as that case is
+    // matched earlier
+    note = "That's emitted if A is not a `String`"
+)]
+#[diagnostic::on_unimplemented(
+    message="message",
+    label="label",
+    note="That's emitted if neither of the condition above are meet",
+)]
+trait MyIterator<A> {
+    fn next(&mut self) -> A;
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+A possible drawback is that this feature adds additional complexity to the compiler implementation. The compiler needs to handle an additional attribute namespace with at least one additional attribute. 
+
+Another drawback is that crates might hint lower quality error messages than the compiler itself. Technically the compiler would be free to ignore such hints, practically I would assume that it is impossible to judge the quality of such error messages in an automated way.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+This proposal tries to improve error messages generated by rustc. It would give crate authors a tool to influence what error message is emitted in a certain situation, as they might sometimes want to provide specific details on certain error conditions. Not implementing this proposal would result in the current status quo. Currently the compiler always shows a "general" error message, even if it would be helpful to show additional details.
+
+There are alternative designs for the proposed `on_unimplemented` attribute:
+
+* The `on()` based filtering might be replaceable by placing the attribute on negative trait impls. This would turn a filter like
+```rust
+#[diagnostic::on_unimplemented(
+    on(Self = `String`, message = "Strings do not implement `IntoIterator` directly")
+)]
+trait IntoIterator {}
+```
+
+into the following negative trait impl:
+```rust
+#[diagnostic::on_unimplemented(message = "Strings do not implement `IntoIterator` directly")]
+impl !IntoIterator for String {}
+```
+
+This would simplify the syntax of the proposed attribute, but in turn block the implementation of type based filtering on the stabilization of `negative_impls`. On the other hand it would likely simplify writing more complex filters, that match only a certain generic set of types and it would prevent "duplicating" the filter-logic as this reuses the exiting trait system. The large disadvantage of this approach is that it couples error messages to the crates public API. Removing a negative trait impl is a breaking change, removing a `#[on_unimplemented]` attribute is only a change in the emitted compiler error.
+
+* Allow `#[diagnostic::on_unimplemented]` to be placed on types instead of traits. This would allow third party crates to customize the error messages emitted for unsatisfied trait bounds with out of crate traits. This feels more like an extension tho the proposed attribute.
```

Comment:
I don't see this as an alternative but more as future work.

Reply:
I'd like to have both (eventually). Current `#[rustc_on_unimplemented]` works well mostly because std is mostly defining the traits, not using traits defined by others.

Reply:
Will move that point to future possibilities :+1: